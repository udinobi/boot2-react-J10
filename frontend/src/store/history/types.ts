import { Action } from "redux"

import { Location, LocationSelectedAction } from "../location/city/types"

export interface HistoryLocation {
    location: Location
}

export interface HistoryState {
    locations: Location[]
}

export enum HistoryActionType {
    LOCATION_RELOAD = "LOCATION_RELOAD",
    LOCATION_REMOVE = "LOCATION_REMOVE"
}

export interface LocationReloadAction extends Action {
    type: HistoryActionType.LOCATION_RELOAD
    payload: {
        location: Location
    }
}

export interface LocationRemoveAction extends Action {
    type: HistoryActionType.LOCATION_REMOVE
    payload: {
        location: Location
    }
}

export type HistoryActions =
    LocationRemoveAction | LocationSelectedAction
