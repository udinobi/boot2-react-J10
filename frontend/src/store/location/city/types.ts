import { Action } from "redux"

import { Option } from "tsoption"

import { Country, CountrySelectedAction } from "../country/types"

export interface Coordinates {
    lat: string
    lon: string
}

export interface Location {
    geoId: number
    name: string
    coord: Coordinates
    country: Country
}

export interface CityState {
    country: Option<Country>,
    suggestions: Location[]
}

export enum CityActionType {
    LOCATION_SELECTED = "LOCATION_SELECTED",
    SUGGESTIONS_FAILURE = "SUGGESTIONS_FAILURE",
    SUGGESTIONS_LOOKUP = "SUGGESTIONS_LOOKUP",
    SUGGESTIONS_RESET = "SUGGESTIONS_RESET",
    SUGGESTIONS_RETRIEVED = "SUGGESTIONS_RETRIEVED"
}

export interface LocationSelectedAction extends Action {
    type: CityActionType.LOCATION_SELECTED
    payload: {
        location: Location
    }
}

export interface SuggestionsFailureAction extends Action {
    type: CityActionType.SUGGESTIONS_FAILURE
}

export interface SuggestionsLookupAction extends Action {
    type: CityActionType.SUGGESTIONS_LOOKUP,
    payload: {
        country: Country,
        locationTerm: string
    }
}

export interface SuggestionsResetAction extends Action {
    type: CityActionType.SUGGESTIONS_RESET
}

export interface SuggestionsRetrievedAction extends Action {
    type: CityActionType.SUGGESTIONS_RETRIEVED
    payload: {
        country: Country
        suggestions: Location[]
    }
}

export type CityActions =
    CountrySelectedAction |
    SuggestionsFailureAction | SuggestionsLookupAction |
    SuggestionsResetAction | SuggestionsRetrievedAction
