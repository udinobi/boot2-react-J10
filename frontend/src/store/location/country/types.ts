import { Action } from "redux"

export interface Country {
    code: string
    name: string
}

export interface CountryState {
    countries: Country[]
}

export enum CountryActionType {
    COUNTRIES_FAILURE = "COUNTRIES_FAILURE",
    COUNTRIES_LOADING = "COUNTRIES_LOADING",
    COUNTRIES_RETRIEVED = "COUNTRIES_RETRIEVED",
    COUNTRY_SELECTED = "COUNTRY_SELECTED"
}

export interface CountriesFailureAction extends Action {
    type: CountryActionType.COUNTRIES_FAILURE
}

export interface CountriesLoadingAction extends Action {
    type: CountryActionType.COUNTRIES_LOADING
}

export interface CountriesRetrievedAction extends Action {
    type: CountryActionType.COUNTRIES_RETRIEVED
    payload: {
        countries: Country[]
    }
}

export interface CountrySelectedAction extends Action {
    type: CountryActionType.COUNTRY_SELECTED
    payload: {
        country: Country
    }
}

export type CountryActions =
    CountriesFailureAction | CountriesLoadingAction | CountriesRetrievedAction
