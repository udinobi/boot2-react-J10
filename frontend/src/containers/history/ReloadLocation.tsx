import { Button } from "antd"
import React from "react"
import { connect } from "react-redux"

import {
    HistoryProps,
    mapHistoryDispatchToProps
} from "../../store/history/actions"

import { HistoryItem } from "../../store/history/types"

export class ReloadLocation extends React.Component<HistoryProps & HistoryItem> {

    constructor(props: HistoryProps & HistoryItem) {
        super(props)
    }

    public render() {
        return (
            <span style={{ marginLeft: 8 }}>
                <Button
                    /* tslint:disable:jsx-boolean-value */
                    data-tip data-for="ReloadLocation"
                    icon="reload"
                    /* tslint:disable:jsx-no-lambda */
                    onClick={() => this.props.reloadLocation(this.props)}
                    shape="circle"
                    size="small"
                    type="primary"
                />
            </span>
        )
    }
}

export default connect(undefined, mapHistoryDispatchToProps)(ReloadLocation);
