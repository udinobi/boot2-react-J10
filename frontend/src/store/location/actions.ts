import { ActionCreator } from "redux"

import { Location } from "./types"

import { LocationChangedAction } from "./types"

export const changeCurrentLocation: ActionCreator<LocationChangedAction> = (location: Location) => ({
    payload: {
        location,
    },
    type: "location/LOCATION_CHANGED"
})
