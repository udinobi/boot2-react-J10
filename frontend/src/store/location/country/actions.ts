import axios, { AxiosResponse } from "axios"
import { ActionCreator, Dispatch } from "redux"

import {
    CountriesFailureAction, CountriesLoadingAction, CountriesRetrievedAction,
    Country, CountryActions, CountryActionType,
    CountrySelectedAction, CountryState
} from "./types"

// CountriesFailureAction --------------------------------------------------------------------------

const countriesFailure: ActionCreator<CountriesFailureAction> = () => ({
    type: CountryActionType.COUNTRIES_FAILURE
})

// CountriesLoadAction -----------------------------------------------------------------------------

const countriesLoading: ActionCreator<CountriesLoadingAction> = () => ({
    type: CountryActionType.COUNTRIES_LOADING
})

// Not checking yet, at this stage in our CSR app, if these properties are set or not.
// In a future SSR rendering feature branch we can also consider to implement logging
// to the server-side RT.
const url =
    `${process.env.REACT_APP_SERVICE_URL}${process.env.REACT_APP_LOAD_COUNTRIES_PATH}`

export const loadCountries = () =>
    async (dispatch: Dispatch<CountryActions>) => {
        dispatch(countriesLoading())

        try {
            const response: AxiosResponse<CountryState> = await axios.get(url)
            dispatch(countriesRetrieved(response.data))
        } catch (error) {
            dispatch(countriesFailure())
        }
    }
        
// CountriesRetrievedAction ------------------------------------------------------------------------

const countriesRetrieved: ActionCreator<CountriesRetrievedAction> = (countries: Country[]) => ({
    payload: { countries },
    type: CountryActionType.COUNTRIES_RETRIEVED,
})

// CountrySelectedAction ---------------------------------------------------------------------------

export const countrySelectedAction: ActionCreator<CountrySelectedAction> = (country: Country) => ({
    payload: { country },
    type: CountryActionType.COUNTRY_SELECTED
})

export const countrySelected = (country: Country) =>
    (dispatch: Dispatch<CountrySelectedAction>) => dispatch(countrySelectedAction(country))

// Action mapping ----------------------------------------------------------------------------------

export interface CountriesProps {
    countries: Country[]
    countrySelected: typeof countrySelected
    loadCountries: typeof loadCountries
}

export const mapCountryDispatchToProps = {
    countrySelected,
    loadCountries
}
    