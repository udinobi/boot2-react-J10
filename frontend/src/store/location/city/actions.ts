import { ActionCreator, Dispatch } from "redux"

import {
    CityActionType,
    Location, LocationSelectedAction,
    SuggestionsFailureAction, SuggestionsLookupAction,
    SuggestionsResetAction, SuggestionsRetrievedAction
} from "./types"

import { Country } from "../country/types"

// LocationSelectedAction --------------------------------------------------------------------------

const locationSelectedAction: ActionCreator<LocationSelectedAction> = (location: Location) => ({
    payload: { location },
    type: CityActionType.LOCATION_SELECTED
})

export const locationSelected = (location: Location) =>
    (dispatch: Dispatch<LocationSelectedAction | SuggestionsResetAction>) => {
        dispatch(suggestionsReset())
        dispatch(locationSelectedAction(location))
    }

// SuggestionsFailureAction ------------------------------------------------------------------------

export const suggestionsFailure: ActionCreator<SuggestionsFailureAction> = () => ({
    type: CityActionType.SUGGESTIONS_FAILURE
})

// SuggestionsLookupAction -------------------------------------------------------------------------

export const suggestionsLookup: ActionCreator<SuggestionsLookupAction> = (country: Country, locationTerm: string) => ({
    payload: { country, locationTerm },
    type: CityActionType.SUGGESTIONS_LOOKUP
})

// SuggestionsResetAction ------------------------------------------------------------------------

const suggestionsReset: ActionCreator<SuggestionsResetAction> = () => ({
    type: CityActionType.SUGGESTIONS_RESET
})

// SuggestionsRetrievedAction ----------------------------------------------------------------------

export const suggestionsRetrieved: ActionCreator<SuggestionsRetrievedAction> = (country: Country, suggestions: Location[]) => ({
    payload: { country, suggestions },
    type: CityActionType.SUGGESTIONS_RETRIEVED
})

// Action mapping ----------------------------------------------------------------------------------

export interface CityDispatchProps {
    locationSelected: typeof locationSelected
    suggestionsLookup: typeof suggestionsLookup
    suggestionsReset: typeof suggestionsReset
}

export const mapCityDispatchToProps = {
    locationSelected,
    suggestionsLookup,
    suggestionsReset
}
