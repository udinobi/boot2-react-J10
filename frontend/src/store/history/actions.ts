import { ActionCreator, Dispatch } from "redux"

import { Location } from "../location/city/types"

import {
    HistoryActionType, HistoryItem, LocationReloadAction, LocationRemoveAction
} from './types'

// LocationReloadAction ----------------------------------------------------------------------------

const locationReloadAction: ActionCreator<LocationReloadAction> = (history: HistoryItem) => ({
    payload: { history },
    type: HistoryActionType.LOCATION_RELOAD
})

export const reloadLocation = (history: HistoryItem) =>
    (dispatch: Dispatch<LocationReloadAction>) => dispatch(locationReloadAction(history))

// LocationRemoveAction ----------------------------------------------------------------------------

const locationRemoveAction: ActionCreator<LocationRemoveAction> = (location: Location) => ({
    payload: { location },
    type: HistoryActionType.LOCATION_REMOVE
})

const removeLocation = (location: Location) =>
    (dispatch: Dispatch<LocationRemoveAction>) => dispatch(locationRemoveAction(location))

// Action mapping ----------------------------------------------------------------------------------

export interface HistoryDispatchProps {
    reloadLocation: typeof reloadLocation
    removeLocation: typeof removeLocation
}

export const mapHistoryDispatchToProps = {
    reloadLocation,
    removeLocation
}
