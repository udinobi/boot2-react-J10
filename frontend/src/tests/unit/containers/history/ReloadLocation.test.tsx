import React from "react"
import * as Renderer from 'react-test-renderer'

import { none } from "ts-option"

import  { ReloadLocation } from "../../../../containers/history/ReloadLocation"

import { HistoryProps } from "../../../../store/history/actions"

import { HistoryItem } from "../../../../store/history/types"

describe("--> Snapshot for Containers > history > 'ReloadLocation'", () => {
    let props: HistoryProps & HistoryItem

    beforeEach(() => {
        props = {
            lastUpdate: none,
            location: {
                coord: { lat: "", lon: "" },
                country: { code : "IT", name: "Italy" },
                geoId: 0,
                name: "",
                tz: ""
            },
            reloadLocation: jest.fn(),
            removeLocation: jest.fn(),
            weatherData: none
        }
    })

    it("matches", () => {
        const component = Renderer.create(
            <ReloadLocation {...props} />    
        )

        const tree = component.toJSON()
        expect(tree).toMatchSnapshot();
    });
})
