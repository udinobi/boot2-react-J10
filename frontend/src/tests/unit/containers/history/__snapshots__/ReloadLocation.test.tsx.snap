// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`--> Snapshot for Containers > history > 'ReloadLocation' matches 1`] = `
<span
  style={
    Object {
      "marginLeft": 8,
    }
  }
>
  <button
    className="ant-btn ant-btn-primary ant-btn-circle ant-btn-sm ant-btn-icon-only"
    data-for="ReloadLocation"
    data-tip={true}
    onClick={[Function]}
    type="button"
  >
    <i
      className="anticon anticon-reload"
    />
  </button>
</span>
`;
